Modifies the loading of the jQuery library by:

WhistlingDonkey opened the pull request:
- Providing further major jQuery CDN's as a choice (there is more than Google)
without having to use the custom field.

- Changes the fallback to the local jQuery source instead of Google.

- Removes potential confusion around the use of 'jquery_auto' by removing it, as
it did nothing but use Google anyway.

** Note, this change replaces the modSetting 'cdn' with 'google_cdn',
'jquery_cdn' and 'microsoft_cdn' so if users have set the jQuery source as
'cdn' (Google) they will fallback to the local source until they resave their
settings. ***

Signed by WhistlingDonkey, github@pyro.life

WhistlingDonkey committed:
Modifies the loading of the jQuery library by:

- Providing further major jQuery CDN's as a choice (there is more than Google)
without having to use the custom field.

- Changes the fallback to the local jQuery source instead of Google.

- Removes potential confusion around the use of 'jquery_auto' by removing it, as
it did nothing but use Google anyway.

** Note, this change replaces the modSetting 'cdn' with 'google_cdn',
'jquery_cdn' and 'microsoft_cdn' so if users have set the jQuery source as
'cdn' (Google) they will fallback to the local source until they resave their
settings. ***

Signed by WhistlingDonkey, github@pyro.life

Sesquipedalian commented on Sources/Load.php:
```diff
@@ -2469,18 +2469,20 @@ function loadTheme($id_theme = 0, $initialize = true)
 	// Add the JQuery library to the list of files to load.
-	if (isset($modSettings['jquery_source']) && $modSettings['jquery_source'] == 'cdn')
-		loadJavaScriptFile('https://ajax.googleapis.com/ajax/libs/jquery/' . JQUERY_VERSION . '/jquery.min.js', array('external' => true, 'seed' => false), 'smf_jquery');
+	$jQueryUrls = array ('google_cdn' => 'https://ajax.googleapis.com/ajax/libs/jquery/'. JQUERY_VERSION . '/jquery.min.js', 'jquery_cdn' => 'https://code.jquery.com/jquery-'. JQUERY_VERSION . '.min.js', 'microsoft_cdn' => 'https://ajax.aspnetcdn.com/ajax/jQuery/jquery-'. JQUERY_VERSION . '.min.js');
```
> ... so if users have set the jQuery source as 'cdn' (Google) they will fallback to the local source until they resave their settings.

To avoid this, please change 'google_cdn' back to just 'cdn'.

If we were starting from scratch, I could see an argument for using jQuery's own CDN as the default rather than Google's, but at this point we should avoid changing the current behaviour unless necessary.

WhistlingDonkey replied:
The same will happen if they have the jquery source set to 'auto' or haven't even set the source by saving the page, they will now fallback to the local source instead of Google which was the previous fallback and I guess it would be advisable not to break that too?

Would adding the line `if (isset($modSettings['jquery_source']) && ($modSettings['jquery_source'] == 'cdn' || $modSettings['jquery_source'] == 'auto')) $modSettings['jquery_source'] = 'google_cdn';`or something along those lines, or simply expanding the array to include cdn and auto be a better idea to capture the old settings?

Sesquipedalian replied:
In this case, the value `auto` means the admin said "I dunno, just use whatever the SMF devs have decided would be a good default", so that's not an issue. But the value `cdn` means that the admin specifically selected "Google CDN" as their jQuery source, so we should maintain that if we can.

... Also, the more I think about it, the more convinced I become that a CDN (I don't really care which one) should be the default, rather than the local file. Using a CDN for this is better both for the server and for the end user, so it should be the default unless the admin has a specific reason not to do so.

Sesquipedalian commented on Sources/Load.php:
```diff
@@ -2469,18 +2469,20 @@ function loadTheme($id_theme = 0, $initialize = true)
+	if (array_key_exists($modSettings['jquery_source'], $jQueryUrls))
+		loadJavaScriptFile($jQueryUrls[$modSettings['jquery_source']], array('external' => true, 'seed' => false), 'smf_jquery');
 
 	elseif (isset($modSettings['jquery_source']) && $modSettings['jquery_source'] == 'local')
 		loadJavaScriptFile('jquery-' . JQUERY_VERSION . '.min.js', array('seed' => false), 'smf_jquery');
```
Since `local` will be the default that we fall back to a few lines below here, this `elseif` block can be removed.

Sesquipedalian replied:
On second thought, please keep this block and instead change the default back to a CDN.

Sesquipedalian commented on Sources/ManageSettings.php:
```diff
@@ -179,9 +179,10 @@ function ModifyBasicSettings($return_config = false)
 			'jquery_source',
 			array(
-				'auto' => $txt['jquery_auto'],
+				'google_cdn' => $txt['jquery_google_cdn'],
```
Please change 'google_cdn' back to just 'cdn'.

WhistlingDonkey replied:
See comment for Load.php

Sesquipedalian commented on Sources/Profile-Export.php:
```diff
@@ -1865,7 +1865,20 @@ function ($a, $b)
 
 	// Next, we need to do for JavaScript what we just did for CSS.
-	loadJavaScriptFile('https://ajax.googleapis.com/ajax/libs/jquery/' . JQUERY_VERSION . '/jquery.min.js', array('external' => true, 'seed' => false), 'smf_jquery');
+	$jQueryUrls = array ('google_cdn' => 'https://ajax.googleapis.com/ajax/libs/jquery/'. JQUERY_VERSION . '/jquery.min.js', 'jquery_cdn' => 'https://code.jquery.com/jquery-'. JQUERY_VERSION . '.min.js', 'microsoft_cdn' => 'https://ajax.aspnetcdn.com/ajax/jQuery/jquery-'. JQUERY_VERSION . '.min.js');
+	
+	if (array_key_exists($modSettings['jquery_source'], $jQueryUrls))
+		loadJavaScriptFile($jQueryUrls[$modSettings['jquery_source']], array('external' => true, 'seed' => false), 'smf_jquery');
+
+	elseif (isset($modSettings['jquery_source']) && $modSettings['jquery_source'] == 'local')
+		loadJavaScriptFile('jquery-' . JQUERY_VERSION . '.min.js', array('seed' => false), 'smf_jquery');
+
+	elseif (isset($modSettings['jquery_source'], $modSettings['jquery_custom']) && $modSettings['jquery_source'] == 'custom')
+		loadJavaScriptFile($modSettings['jquery_custom'], array('external' => true, 'seed' => false), 'smf_jquery');
+
+	// Fall back to the local source
+	else
+		loadJavaScriptFile('jquery-' . JQUERY_VERSION . '.min.js', array('seed' => false), 'smf_jquery');
```
Please revert these changes in Profile-Export.php.

In this case, the jQuery library *always* needs to be loaded from a reliable CDN. This is because the export files are meant to function as permanent, standalone files that will continue to work as expected even if the forum they were exported from disappears one day.

WhistlingDonkey replied:
I can't agree that Google is in anyway more reliable than the other CDN's, or even the local copy since the page is served from the same place, Google are just as fallable as the others and a user would likely have the chosen CDN's copy in cache by the time they land on this page as well but if it needs to be changed back then it shall be done, there is probably more to this one than I know.

Sesquipedalian replied:
> ... or even the local copy since the page is served from the same place...

This seems to be the source of the misunderstanding. ProfileExport.php doesn't generate dynamic pages to be served to the browser. It generates static HTML files that the user can download to their own hard drive for archival purposes. Therefore, _everything_ in these files must be either (a) served from localhost, (b) embedded directly in the HTML file, or (c) served from a very reliable external source, where "reliable" means "highly unlikely to succumb to link rot any time in the foreseeable future"—in other words, from a major CDN. So yes, this really does need to be changed back.

Sesquipedalian commented:
Also, please open an issue explaining why these changes are required (or at least recommended).

WhistlingDonkey commented:
#6913

Sesquipedalian commented on Sources/Load.php:
```diff
@@ -2469,18 +2469,20 @@ function loadTheme($id_theme = 0, $initialize = true)
 	elseif (isset($modSettings['jquery_source'], $modSettings['jquery_custom']) && $modSettings['jquery_source'] == 'custom')
 		loadJavaScriptFile($modSettings['jquery_custom'], array('external' => true, 'seed' => false), 'smf_jquery');
 
-	// Auto loading? template_javascript() will take care of the local half of this.
+	// Fall back to the local source
 	else
-		loadJavaScriptFile('https://ajax.googleapis.com/ajax/libs/jquery/' . JQUERY_VERSION . '/jquery.min.js', array('external' => true, 'seed' => false), 'smf_jquery');
+		loadJavaScriptFile('jquery-' . JQUERY_VERSION . '.min.js', array('seed' => false), 'smf_jquery');
```
Looking at this again, I think it is best to keep the default as it was. In the vast, vast majority of cases, using a CDN is the right choice here.

WhistlingDonkey committed:
Applies requested changes.

- Reverts Profile-Export.php to original code.
- Reverts the fallback source to Google.
- Reverts 'google_cdn' setting to 'cdn' so as not to break current settings.
- Reintroduces 'auto' setting and selection option so as not to break current settings.
- Renames 'Auto' that admins used to see for the 'auto' selection option to
'Forum default' to remove any ambiguity or confusion on the use of the work auto.
- Updated help file to put in the 'Forum default' setting.

Signed by WhistlingDonkey github@pyro.life

Sesquipedalian commented on Sources/ManageSettings.php:
```diff
@@ -179,9 +179,11 @@ function ModifyBasicSettings($return_config = false)
 			'jquery_source',
 			array(
-				'auto' => $txt['jquery_auto'],
+				'auto' => $txt['jquery_forum_default'],
```
```suggestion
```

Sesquipedalian commented on Themes/default/languages/Help.english.php:
```diff
@@ -265,7 +265,7 @@
 $helptxt['enable_ajax_alerts'] = 'This option allows your members to receive AJAX notifications. This means that members don\'t need to refresh the page to get new notifications.<br><strong>DO NOTE:</strong> This option might cause a severe load at your server with many users online.';
 $helptxt['alerts_auto_purge'] = 'Once an alert has been read, it is rarely needed again. For performance reasons, it is a good idea to automatically delete them after a while.';
-$helptxt['jquery_source'] = 'This will determine the source used to load the jQuery Library. <em>Auto</em> will use the CDN first and if not available fall back to the local source. <em>Local</em> will only use the local source. <em>CDN</em> will only load it from Google CDN network';
+$helptxt['jquery_source'] = 'This will determine the source used to load the jQuery Library. <em>Forum default</em> uses the forum default source, <em>Google CDN, jQuery CDN</em> and <em>Microsoft CDN</em> will load the jQuery library from those respective CDN networks. <em>Local</em> will only use the local source. <em>Custom</em> allows you to specify a custom URL for the library.';
```
```suggestion
$helptxt['jquery_source'] = 'This will determine the source used to load the jQuery Library. <em>Google CDN</em>, <em>jQuery CDN</em>, and <em>Microsoft CDN</em> will load the jQuery library from those respective CDN networks. <em>Local</em> will only use the local source. <em>Custom</em> allows you to specify a custom URL for the library.';
```

Sesquipedalian commented on Themes/default/languages/ManageSettings.english.php:
```diff
@@ -72,8 +72,10 @@
-$txt['jquery_cdn'] = 'Google CDN';
-$txt['jquery_auto'] = 'Auto';
+$txt['jquery_google_cdn'] = 'Google CDN';
+$txt['jquery_forum_default'] = 'Forum default';
```
```suggestion
```

Sesquipedalian reviewed (changes requested):
Looking pretty good.

At this point, we can remove the "auto" option entirely. Forums that currently have "auto" selected will end up using the Google CDN, just as they do now, and "Google CDN" will appear as the default option in the select menu for them when they load the admin page. In light of that, we can just ditch "auto" and move on without it.

WhistlingDonkey committed:
Applies requested changes

- Removes references to 'auto' and 'Forum default'. The first and default
option presented to admins is now Google CDN.

Signed by WhistlingDonkey github@pyro.life

WhistlingDonkey commented:
I had reintroduced auto as 'Forum default' as I figured it'd serve as a reminder that the admin hadn't explicitly set a source but with Google CDN as the first and default choice it is not needed. Requested changes applied.

Sesquipedalian commented:
Thanks for contributing!
